### PR TITLE
Updating debugprotocol for simulator

### PR DESCRIPTION
**Pull request by riknoll**
This PR updates the debug protocol that the simulator uses to something a bit more usable. This is still very much a work in progress.

### CODE REVIEW

File: webapp/public/run.html [hunk 1]
```diff
@@ -84,19 +84,22 @@
...
+        var debugSim = !!/debugSim=1/i.exec(window.location.href);
+        var sims = document.getElementById('simulators');
+
+        if (!id && !code && !debugSim) {
             console.error("missing id or code");
             loading.innerText = 'Oops, wrong arguments...';
             return;
         }
-        var options = { id: id ? id[1] : undefined, code: code ? decodeURIComponent(code[1]) : undefined };
-        var sims = document.getElementById('simulators');
```
Comment (abchatra) [inline webapp/public/run.html]: Duplicate code? It is on line 95 as well.

**Comment by abchatra** [comment]
This PR looks good to me.

File: cli/cli.ts [hunk 2]
```diff
@@ -3131,7 +3148,13 @@ function stringifyTranslations(strings: pxt.Map<string>): string {
 export function buildAsync(parsed: commandParser.ParsedCommand) {
     forceCloudBuild = !!parsed.flags["cloud"];
 
-    return buildCoreAsync({ mode: BuildOption.JustBuild })
+    let mode = BuildOption.JustBuild;
+    if (parsed.flags["debug"]) {
+        mode = BuildOption.DebugSim;
+    }
+
+
```
Comment (pelikhan) [inline cli/cli.ts]: spurious line

File: pxtrunner/debugRunner.ts [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+
+    /**
+     * Message containing code and debug information for simulator
+     */
+    interface RunCodeMessage extends RunnerMessage {
+        subtype: "runcode";
+        code: string;
+        usedParts: string[];
+        usedArguments: { [index: string]: string };
```
Comment (pelikhan) [inline pxtrunner/debugRunner.ts]: we have a map type for those

File: pxtrunner/debugRunner.ts [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+        public onError(cb: (e?: any) =>  void): void {
+            this.errorListener = cb;
+        }
+
+        public onClose(cb: () => void): void {
+            this.closeListener = cb;
+        }
+
+        public close(): void {
+            // TODO
```
Comment (pelikhan) [inline pxtrunner/debugRunner.ts]: Next PR?

File: pxtrunner/debugRunner.ts [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+                    source: {
+                        path: bp.fileName
+                    }
+                }]);
+            });
+
+            this.session.runCode(msg.code, msg.usedParts, msg.usedArguments, new pxsim.BreakpointMap(breakpoints), pxt.appTarget.simulator.boardDefinition);
+        }
+
+        private sendRunnerMessage(subtype: string, msg: {[index: string]: string} = {}) {
```
Comment (pelikhan) [inline pxtrunner/debugRunner.ts]: use Map

File: pxtsim/runtime.ts [hunk 6]
```diff
@@ -384,6 +384,9 @@ namespace pxsim {
                         case "stepinto":
                             breakAlways = true
                             break
+                        case "stepout":
+                            breakAlways = true;
+                            breakFrame = s.parent || s;
```
Comment (pelikhan) [inline pxtsim/runtime.ts]: missing break (i know it's the last case but still)

File: cli/cli.ts [hunk 7]
```diff
@@ -2933,16 +2934,32 @@ function buildCoreAsync(buildOpts: BuildCoreOptions): Promise<pxtc.CompileOption
...
         })
         .then((res): Promise<void | pxtc.CompileOptions> => {
-            U.iterMap(res.outfiles, (fn, c) =>
-                mainPkg.host().writeFile(mainPkg, "built/" + fn, c))
+            U.iterMap(res.outfiles, (fn, c) => {
+                if (fn !== pxtc.BINARY_JS) {
+                    mainPkg.host().writeFile(mainPkg, "built/" + fn, c);
+                }
+                else {
+                    mainPkg.host().writeFile(mainPkg, "built/debug/" + fn, c);
```
Comment (riknoll) [inline cli/cli.ts]: This is a CLI-only flag right now, but you're right that we haven't really had any nested files like this before.

**Pull request merged**